docs/source/guides/testing-react-apollo.md

```diff
@@ -0,0 +1,334 @@
+  renderer.create(<Dog name="Buck" />);
+});
+```
+
+If we ran this test, we would get an error because Apollo Client isn't available on the context for the `Query` component to consume.
+
+In order to fix this we could wrap the component in an `ApolloProvider` and pass an instance of Apollo Client to the `client` prop. However, this will cause our tests to run against an actual backend which makes the tests very unpredictable for the following reasons:
+
+- The server could be down
+- There may be no network connection
+- The results are not guaranteed to be the same for every query
+
+```js
+// Not predictable
+it('renders without error', () => {
+  renderer.create(
+    <ApolloProvider client={client}>
+      <Dog name="Buck" />
+    </ApolloProvider>,
+  );
+});
+```
+
+## MockedProvider
+
+To test the component in true isolation, we can mock all calls to the GraphQL endpoint. This makes the UI consistent every time tests are run, since they don't depend on any remote data.
+
+React Apollo provides the `MockedProvider` component in the `react-apollo/test-utils` library to do just that! `MockedProvider` allows us to specify the exact results that should be returned for a certain query using the `mocks` prop.
+
+Here's an example of a test for the above `Dog` component using `MockedProvider`:
+
+```js
+// dog.test.js
+
+import { MockedProvider } from 'react-apollo/test-utils';
+
+// The component AND the query need to be exported
+import { GET_DOG_QUERY, Dog } from './dog';
+
+const mocks = [
+  {
+    request: {
+      query: GET_DOG_QUERY,
+      variables: {
+        name: 'Buck',
+      },
+    },
+    result: {
+      data: {
+        dog: { id: '1', name: 'Buck', breed: 'bulldog' },
+      },
+    },
+  },
+];
+
+it('renders without error', () => {
+  renderer.create(
+    <MockedProvider mocks={mocks} addTypename={false}>
+      <Dog name="Buck" />
+    </MockedProvider>,
+  );
+});
+```
+
+This example shows how to define the mocked response for a query. The `mocks` array takes objects with specific `request`s and their associated `result`s. In this example, when the provider receives a `GET_DOG_QUERY` with the specified variables, it returns the object under the `result` key.
+
+### addTypename
+
+You may notice the prop being passed to the `MockedProvider` called `addTypename`. The reason this is here is because of how Apollo Client normally works. When a request is made with Apollo Client normally, it adds a `__typename` field to every object type requested. This is to make sure that Apollo Client's cache knows how to normalize and store the response. When we're making our mocks, though, we're importing the raw queries _without typenames_ from the component files.
+
+If we don't disable the adding of typenames to queries, our imported query won't match the query actually being run by the component during our tests.
+
+> In short, If you don't have `__typename`s in your queries, pass `addTypename={false}` to your `MockedResolver`s.
```

Comment:
I am wondering what the `addTypename` prop is actually used for. When testing I simply want to link some data to the imported query without having to deal with typenames like so:

```js
import { Query } from “./some_component”;

const mocks = [{
    request: {
         query: Query
    },
    result: someData
}]

renderer.create(
<MockedProvider mocks={mocks}>
   <Component />
</MockedProvider>
);
```

I actually created a PR in the past where I simplified the `MockedProvider` (see [#1882](https://github.com/apollographql/react-apollo/pull/1882)) and removed the `addTypename` prop. Back then I created this PR because I was constantly running into the issue that the MockedProvider was not returning data for my query (this was because `addTypename` was true by default). 

Furthermore, I have a [PR]](https://github.com/apollographql/react-apollo/pull/1995) open that sets the `addTypename` prop to false by default because this will allow developers to simply create mock data for their queries without having to worry about typenames.

Is there a specific use case that the `addTypename` prop enables? I have personally never had to use it in my projects and it has actually gotten in my way when testing apollo components. So I am wondering if more users will run into this.

Given that the MockedProvider has not been documented yet, I think that this may be a good moment to flesh out this API before it becomes public knowledge. Thoughts?

Reply:
I think I agree that `false` is a more sensible default.

I think we left it to `true` by default to make sure that it wouldn't be a breaking change to anyone using it currently. We hadn't documented it, but I know there are some people that we had pointed to this code initially, and may be using it in their projects.

Thoughts @jbaxleyiii?

Reply:
Oh ok, if people are actively using it then a breaking change may not be such a good idea. With the future release of react-apollo v3 then perhaps this behavior can be changed. Thanks for clarifying:)

@jbaxleyiii  Do you know by any chance why the `addTypename` prop was actually added in the first place? I'm wondering what use cases it enables

Reply:
Yeah, v3 may be a good place to reconsider this. 

Originally, the mocked provider acted like the normal provider did, and added typename by default. We added this prop to disable that